### PR TITLE
[FLINK-25572][connectors/filesystem] Update File Sink to use decomposed interfaces.

### PR DESCRIPTION
**Pull request by pltbkd**
## What is the purpose of the change

This pull request migrates the File Sink to the new decomposed sink interface.

## Brief change log

 - Refactor FileSink, its components and its tests with the new sink interface. No feature is changed.

## Verifying this change

This change is already covered by existing tests for the origin FileSink.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/committer/FileCommitter.java [hunk 1]
```diff
@@ -48,9 +47,10 @@ public FileCommitter(BucketWriter<?, ?> bucketWriter) {
     }
 
     @Override
-    public List<FileSinkCommittable> commit(List<FileSinkCommittable> committables)
-            throws IOException {
-        for (FileSinkCommittable committable : committables) {
+    public void commit(Collection<CommitRequest<FileSinkCommittable>> requests)
+            throws IOException, InterruptedException {
+        for (CommitRequest<FileSinkCommittable> request : requests) {
```
Comment (gaoyunhaii) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/committer/FileCommitter.java:52]: Perhaps we could handle the exceptions with `request.signalFailedWithKnownReason(e)` and other methods if it throws exception, or we create a follow-up issue to optimize the behavior of committing.

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/writer/FileWriter.java [hunk 2]
```diff
@@ -206,7 +219,7 @@ public void write(IN element, Context context) throws IOException {
             if (!entry.getValue().isActive()) {
                 activeBucketIt.remove();
             } else {
-                committables.addAll(entry.getValue().prepareCommit(flush));
+                committables.addAll(entry.getValue().prepareCommit(endOfInput));
```
Comment (gaoyunhaii) [inline flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/writer/FileWriter.java:222]: Might also change the parameter name of prepareCommit(boolean) to `endOfInput`

**Review by gaoyunhaii — COMMENTED** [review]
Very thanks @pltbkd for drafting the PR! LGTM and I have some small issues.

Also, there are some compile issues in flink-table~

File: flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/committer/FileCommitter.java (hunk 1 above)
Comment (pltbkd) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/committer/FileCommitter.java:52]: This is inherited from the interface. Maybe we can discuss whether to allow throwing exceptions  here or to make a agreement that all exceptions while committing should be reported via the CommitRequest (unless the exception is not bounded to a specific request)?

Comment (gaoyunhaii) [reply flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/committer/FileCommitter.java:52]: I think the key is that we might need to know the result of a single request. We might first not change the behavior here and open a follow-up issue for this part if there are still controversy~

**Comment by pltbkd** [comment]
@gaoyunhaii
Thanks for reviewing. All these issues should have been solved, including the compile issues.

I found that sinkTo(sink2.Sink) is not added in the scala API (DataStream.scala), which results into compiling error in flink-examples. I have added it to resolve the issues, though I'm not should if this should be included in this pr. Please check when you review the new changes. Thanks.

**Review by gaoyunhaii — APPROVED** [review]

**Comment by gaoyunhaii** [comment]
Thanks @pltbkd for the updating! LGTM. 

For the fix I think it could be put as a hot fix in a separate commit in this PR. I'll change it when merging.

**Pull request closed without merging**